include/vara/Solver/ConfigurationFactory.h

```diff
@@ -17,7 +17,7 @@ class ConfigurationIterable {
 public:
   class ConfigurationIterator {
   public:
-    using iterator_category = std::forward_iterator_tag;
+    using iterator_category = std::input_iterator_tag;
```

Comment:
Why can the config iterator no longer be a forward_iterator? Having it at least forward would probably be good to use it in standard algorithms.

Reply:
As far as I understood this, a `forward_iterator` allows to iterate multiple times.
That was never possible with the implementation because incrementing the iterator changes the solver state in a non-reversible way. With the current implementation, it was incorrect to use `forward_iterator` in the first place.

Reply:
At least the following requirements of a `forward_iterator` are violated:
Given `a` and `b`, dereferenceable iterators of type `It`:
- If `a` and `b` compare equal (`a == b` is contextually convertible to `true`) then either they are both non-dereferenceable or `*a` and `*b` are references bound to the same object.
- `a == b` implies `++a == ++b`

For the first point, it would only be true that `*a` and `*b` describe the same configuration, but they are not the same object (they are  `unique_ptr`s pointing to a `Configuration`). In fact, such an object is created every time the iterator is dereferenced or incremented.

Reply:
Thinking more about this, I'm not sure if this could also be an issue for the `input_iterator` specification as it says:

If `a` == `b` and `(a, b)` is in the domain of `==` then `*a` is equivalent to `*b`.

This would only be fulfilled if _equivalent_ refers to conceptual equivalence of the objects (i.e., they represent the same ting/configuration). There is currently no equivalence implemented on `Configuration`s.

Reply:
Hmm, I see. Then let's keep the inputput tag. PS: I don't think they need to be the same object, so input_iterator should be fine.